### DISCUSSION TITLE
BlobModule API

### DISCUSSION THREAD

**Discussion by Wondertan** [celestiaorg/celestia-node#1829] [General]
``` go
// Blob represents any application-specific binary data that anyone can submit to Celestia.
type Blob struct {
	Namespace namespace.ID
	Data []byte
}

// Proof is a collection of nmt.Proofs that verifies the inclusion of the data.
type Proof []nmt.Proof

// Commitment is a Merkle Root of the subtree built from shares of the Blob.
type Commitment []byte

// Commitment computes the commitment of the Blob. 
// It is computed by splitting the blob into shares and building the Merkle subtree to be included after Submit.
func (b *Blob) Commitment() Commitment

// BlobModule provides all necessary
type BlobModule interface {
	// Submit sends PFB transaction and reports the height it landed on.
	// Allows sending multiple Blobs atomically synchronously.
	// Uses default wallet registered on the Node.
	Submit(context.Context, Blob...) (height uint64, error)
	
	// Get retrieves the blob by commitment. 
	Get(context.Context, uint64, namespace.ID, Commitment) (Blob, error)
	
	// GetAll retrieves all the blobs for given namespaces at the given height. 
	GetAll(context.Context, height uint64, namespace.ID...) ([]Blob, error)

    // GetProof gets the Proof for a blob if it exists at the given height under the namespace. 
    GetProof(context.Context, height uint64, namespace.ID, Commitment) (Proof, error)

	// Included checks whether a blob's given commitment(Merkle subtree root) is included at the given height under a particular namespace.
	// Useful for light clients wishing to check the inclusion of the data.
	Included(context.Context, height uint64, namespace.ID, Commitment, Proof) (bool, error)
}
```

Two uncovered/undecided features:
* Exposing inclusion proofs from the API. I am very hesitant to do this for reasons I will explain in a diff thread.
* Inclusion checks for data at a particular share index rather than the whole blob.
    
  cc @musalbas, @liamsi, @tzdybal, @nashqueue

**Comment by musalbas** [comment]
It might be helpful to add the index of the Blob when you get them, in case people need to link them to a PFB that was published (eg to link them to a certain transaction submitter).

**Reply by Wondertan** [reply]
How does the order of those is formed?

**Reply by Wondertan** [reply]
I get what you mean, but our current Shrex stack does not enable requesting a blob by index. Currently, users can request all the blobs and get the one they need by index. This doesn't mean we shouldn't extend shrexnd to enable this at some point

**Reply by musalbas** [reply]
Adding the indexes to []Blob might still be helpful, regardless of whether you can request an individual Blob by index. Otherwise, devs might not be able to link specific blobs to a specific PFB.

It might require a blob response to have a different struct that submitted blog (eg IncludedBlob that extends Blob, vs Blob), where IncludedBlob could also have the proof, for example...

@preston-evans98 may have some thoughts here as they parse PFBs in their rollup.

**Reply by Wondertan** [reply]
> Adding the indexes to []Blob might still be helpful

I am not sure what you mean by this. `[]Blob` is a slice and they have indices

**Reply by musalbas** [reply]
I mean the index of the share inside the data root, not in the slice.

**Reply by Wondertan** [reply]
Ok, we can guarantee 1:1 parity between the order inside the data root and the slice returned by Get. This is what I meant with
> Currently, users can request all the blobs and get the one they need by index.

And I assume users cannot select the index when the Submit PFB, right?

**Reply by musalbas** [reply]
>Ok, we can guarantee 1:1 parity between the order inside the data root and the slice returned by Get. This is what I meant with

That still doesn't allow you to know the index of the blob inside the data root though, as you don't know the index of the first share in the slice inside the data root.

>And I assume users cannot select the index when the Submit PFB, right?

No, that's determined by block producers.

**Reply by preston-evans98** [reply]
Being able to easily identify the blob's location inside of the data root would simplify our node implementation, but I don't think it changes the amount of work we have to do in-circuit - so it's not a huge deal either way.

**Reply by preston-evans98** [reply]
To expand on this a little - IIRC, our current implementation just reconstructs the entire data square outside of the circuit. Since we already have the complete data, it's easy for us to find the index of a particular blob in the data root. And the verifier implicitly checks that index by verifying the merkle proof of inclusion.


But now that I think about it, isn't the index already included in the merkle proof?

**Reply by musalbas** [reply]
>But now that I think about it, isn't the index already included in the merkle proof?

That should be true, yes, if the API includes proofs.

**Reply by mztacat** [reply]
indexing the BlobModule would be great. i find it really hard navigating here

**Comment by Wondertan** [comment]
Also, there is one feature request from rollkit that I want to understand fully: the need to manage blob inclusion on the rollkit side rather than delegating it to celestia-node. By manage, I mean handling networking and verifying them. On the other side, I am proposing to hide/abstract away most of the proof-related logic as an implementation detail of Celestia Node, as we do for DAHeader syncing and sampling.
The API looks like https://github.com/celestiaorg/celestia-node/discussions/1829#discussion-4903368

Pros:

* Less work for Rollkit and other teams. Fetching and proof verification are done by cel-node
* Code for this is implemented in a single place, enabling better maintainability and easier audits. Having this all in an open place benefits our ecosystem.
* We can enable giving proofs in special cases, but the default is celestia-node handling this.
* Enables minimal future and fool proof API for Blobs.
	* Future proof because abstracts away proving system enabling this API to survive changes once Celestia migrates to KZG(not a fact)
	* Fool proof because it hided away complexity
* It does not increase development on Celestia-node team, we anyway have to have such features.

Cons:
* Technically increases networking load on celestia-node p2p network from rollkits, but
	* It's only proofs, not the data
	* In any case, it is accessible to group nodes interested in a particular namespace, so in practice, these proofs can be offloaded from cel-nodes hosted by celestia to cel-node hosted by rollkit network by discovery.
* Anything else you have in mind that I missed

**Reply by liamsi** [reply]
would it be possible to expose both versions? I think in some cases users are fine with node handling the proofs and in some cases users will prefer to handle them themselves. 

I think making a version where the proof nodes are explicitly collected and returned on submit, is probably not too difficult anyways, right? (same for the inclusion check)

**Reply by preston-evans98** [reply]
For zk-rollups, we need to be able to verify blob inclusion proofs in-circuit. So I'd strongly recommend against treating proofs as an internal implementation detail. 

+1 to @liamsi's suggestion of exposing two versions.

**Reply by nashqueue** [reply]
This would require celestia - node to understand which rollkit nodes are connected. I don't think celestia-node should understand the network topology of rollkit. 

Rollkit cannot expect that other DA-Layers will be able to handle proofs for them, so we need to implement gossiping on our side anyways and celestia will be a special case for rollkit.

If rollkit handles the proof, you will get a bestcase where we only request the proof once and gossip over our p2p. 
If celestia node handles the proof every rollkit node will need to request the proof. If we make it waiting and not by request then Celestia node needs to understand rollkit network layer to be able to gossip efficiently to only the people that need it, i.e. rollkit nodes register at celestia level. 

If the rule is imposed that we register by namespace then rollkit is forced to comply limiting some usecases. We might want to use multiple namespaces for different things. 

I feel DA-Interface is already future proof so KZG - commitments should be handled by it. 

There is also a 1-n relation ship between celestia-node and rollkit node, as you can have many rollkit nodes getting served by the same celestia node.

**Reply by Wondertan** [reply]
> would it be possible to expose both versions? I think in some cases users are fine with node handling the proofs and in some cases users will prefer to handle them themselves.

@liamsi, Yes! We can do it.

> I think making a version where the proof nodes are explicitly collected and returned on submit, is probably not too difficult anyways, right? (same for the inclusion check)

It's not difficult and even simpler for celestia-node to offload proofs business to layers above. To keep both versions, we would probably just need to separate out `GetProof` method that enables getting proof for those who need it. The option to return proofs on `Submit` should work, but if we also need to return it from `Available` and there is a use case just to access proof independently of submission and inclusion check, then a separate `GetProof` make more sense to me.

**Reply by Wondertan** [reply]
> For zk-rollups, we need to be able to verify blob inclusion proofs in-circuit. So I'd strongly recommend against treating proofs as an internal implementation detail.

@preston-evans98, Do you need to access the proof only after submission or also during the inclusion check? Do you need proofs outside of submission and inclusion check operations?

**Reply by Wondertan** [reply]
Ok, so we synchronously discussed this in-depth with @nashqueue. We both agreed that delegating proofs(and even data) to the Node software and p2p network has multiple benefits(some of them mentioned in the message). We thoroughly analyzed best-case and worst-case scenarios, including DOS vectors and the Woods attack of the proposal. We came up with a scalable design that keeps the mentioned benefits. In the end, we made a pragmatic decision.Here is the design for the Rollkit(or any other project) that can leverage celestia-node's p2p network for blob and proofs gossiping:
* Node peers have an affinity based on the namespace set they are interested in, which ___implicitly___ follows a Rollkit network topology.
	* Besides being connected to the DA network, peers interested in a unique set of namespaces form unstructured sub-networks.
	* This is achievable via peer discovery by default. The subnetwork maintainers may explicitly apply custom structuring to the network using Node's P2P API.
* Due to the Woods attack vector, peers in the subnetwork ___cannot___ ___follow___ the namespace set by downloading ___all___ the proofs(and potentially data) in the namespace set.
	* Instead, the subnetwork peers gossip with each other notifications about new blob inclusion in the namespace set. A notification represents a content addressable unique commitment to a blob.
		* The notification is only emitted by a peer to its immediate peers in the subnetwork if the blob is included, fetched, and ready to be served further by the peer.
		* This design is inspired by the node's BlockSync and is already implemented. The namespace set affinity is FN discovery, and notifications are made via data root, which is a commitment to the whole square rather than a blob. Extending existing implementation to support namespace subnetworks is a low to medium-effort task.
* The Rollkit header in the best-case soft-commitment scenario has reached the subnetwork peers(including Rollkit LNs) earlier than the blob is included and available. The header contains the commitment to the blob awaiting inclusion, so a Rollkit LN will request the blob by the commitment from the connected DA LN before it's available. Once the notification for the blob is received, the DA LN asks for it from the subnetwork and provides the blob back to the Rollkit node.

However, even though the proposed design is desirable, it requires efforts from celestia-node side which will only be able to allocate resources for this after mainnet. At the same time, Rollkit needs an efficient way to access proof and data asap, so the decision is to implement a custom and single-purpose solution on the Rollkit side to later be swapped out with general purpose solution with the design above powered by Celestia node software stack.

**Reply by musalbas** [reply]
Regardless of how much is internalized in celestia-node, zk rollups will still _always_ need to verify the NMT proofs in-circuit, so celestia-node will still need to expose the raw proofs so that they can be passed into the zk program, in any short- or long- term proposal.

"implement a custom and single-purpose solution on the Rollkit side" does not seem to meet that goal, since that may be fine for Rollkit, but it's not fine for other rollup frameworks such as https://github.com/Sovereign-Labs/Jupiter?

**Reply by Wondertan** [reply]
Celestia node, anyway, must be able to serve proofs, even when Rollkit handles networking independently because the sequencer has to get it from somewhere to share with his network. Jupiter will use the same way to request proofs; it just won't be as efficient as the custom networking that Rollkit aims for now.

**Reply by Wondertan** [reply]
The other question I have to Soverign design. Do they need blob inclusion proofs specifically, e.g spanned over multiple rows, or proofs over just rows is enough for them?

**Reply by preston-evans98** [reply]
> Do you need to access the proof only after submission or also during the inclusion check? Do you need proofs outside of submission and inclusion check operations?

If I understand your question correctly, we just needs proofs during the inclusion check. 

To be more explicit, we make a SNARK proof for each Celestia block that says "I have read all of the blobs from Namespace X". In order to do that, we have the prover download all of the blobs (for the namespace) from Celestia Node and provide them as inputs to the circuit. Then, the circuit checks an NMT proof that (1) all of the blobs "real" (meaning that they are committed to in the DAH) and (2) the set of blobs is "complete" (meaning that the prover didn't leave out any blobs).

**Reply by preston-evans98** [reply]
> The other question I have to Soverign design. Do they need blob inclusion proofs specifically, e.g spanned over multiple rows, or proofs over just rows is enough for them?

We need to verify statements about blobs specifically

**Comment by Wondertan** [comment]
The alternative API which forces users to handle Proofs:

(only affected methods are shown)
```go
	// Awaits inclusion and provides users proofs for the user(Rollkit) to gossip over their network
	Submit(context.Context, Blob...) (height uint64, []Proof, error)
	
	// Only find the correct DAHeader by height and perform an inclusion check against it(usefull for Rollkit Light clients)
	Available(context.Context, height uint64, namespace.ID, hash.Hash, Proof) (bool, error)
```

**Comment by liamsi** [comment]
API looks like a great start to me!

> // Commit computes the hash commitment of the Blob.
func (b *Blob) Commit() hash.Hash


> // Available checks whether the given hash commitment of a blob is available/included at the given height under a particular namespace.
>	// Useful for light clients wishing to check the inclusion of the data.
>	Available(context.Context, height uint64, namespace.ID, hash.Hash) (bool, error)


I think in both cases it remains a bit too vague what commit refers to. Is it the hash of the blob (probs not) is it the same substree root included in the pfb?

**Reply by Wondertan** [reply]
It's a non-interactive subtree root, yes. Will clarify that

**Reply by Wondertan** [reply]
Clarified

**Reply by rootulp** [reply]
```golang
// Commit computes hash commitment of the Blob.
func (b *Blob) Commit() []byte
```

Proposal to use commitment for consistency https://github.com/celestiaorg/celestia-app/issues/1202. 

>  is it the same substree root included in the pfb?

If this commitment refers to the [`ShareCommitments`](https://github.com/rootulp/celestia-app/blob/d2eb6b0fd9f2b83859499ac315962ea7a67bd13f/x/blob/types/tx.pb.go#L37-L38) that a user generates and includes in their PFB transaction then it is not a singular subtree root. Instead it is a Merkle root over multiple subtree roots in the data square. Ref:
1. https://github.com/celestiaorg/celestia-app/tree/main/x/blob#how-is-the-sharecommitment-generated
2. https://github.com/rootulp/celestia-app/blob/d2eb6b0fd9f2b83859499ac315962ea7a67bd13f/x/blob/types/payforblob.go#L127-L133

**Comment by nashqueue** [comment]
Rollkit will need an inclusion check of arbitrary shares that might not convert into a commitment. Let me give you the example:

[S1][S2][S3][S4][S5][S6][S7][S8] This Blob is 8 Shares long and and you can generate a PFB commitment over it with ease. 

Now lets say a full node found fraud in those shares and wants to share it with the LN. They will need to share a blob inclusion proof to the shares those specific shares. 

Those shares might be in the middle of the block. Because of non-imnteractive default rules you cannot have a pfb commitment over those. 

Lets say the fraudulent shares are in [S2][S3]

The merkle tree over the shares looks like this: 
   /    \       /    \       /    \       /    \    
[S1][S2][S3][S4][S5][S6][S7][S8]

Thats why there is no PFB commitment over 
 \       /  
[S2][S3]

One way to reference those 2 Shares is by (height,index,length) and not by commitment.

**Reply by rootulp** [reply]
> Now lets say a full node found fraud in those shares and wants to share it with the LN.

Is this assuming that the roll-up LN does not have large enough hardware resource requirements to accept a blob inclusion proof that provides S1 through S8?

**Reply by nashqueue** [reply]
If you allow fraud proofs to be S1 to S8 then a full node can send a fraud proof to a LN every block . They will have to download the whole block which in turn will make them a full node. That is why you need this scheme to only give a portion of the data.

**Reply by rootulp** [reply]
> then a full node can send a fraud proof to a LN every block

If the fraud proof is valid, I assume the light node will halt after verifying the first fraud proof. If the fraud proof is invalid, I assume the LN will decrease the peer score of the FN that sent this fraud proof or blocklist it entirely.

Under what circumstances do we expect a LN to receive a **valid** fraud proof for several sequential blocks?

**Reply by nashqueue** [reply]
S1-S8 is one Block. The compact fraud proof would only require the LN to get a small portion of the block. You only know if the fraud proof is invalid after you check the last ISR (in the fraud-proof). This means that you have to execute all Txs to be able to know if the last ISR is correct or not.

**Reply by rootulp** [reply]
> This means that you have to execute all Txs to be able to know if the last ISR is correct or not.

In this context I think "you" is referring to the LN. If so, I don't see why this is problematic unless there is an assumption that the LN doesn't have hardware resource requirements to download S1 through S8. Is that the case? 

> One way to reference those 2 Shares is by (height,index,length) and not by commitment.

An alternative may be to provide an NMT inclusion proof to the subtree root that contain the shares with fraud. In the example I think a FN could convince a LN that S1-S4 exist in the NMT inner node b. The LN would have to download the shares S1-S4. Then I think other NMT proofs can be provided (with hashes and not the raw shares) to prove that S1-S4 is part of the blob S1-S8 and that this blob exists in the data square.

```ascii
                a
              /   \
            /       \
          /           \
        b              c
      /   \         /     \
    d      e       f       g
  /  \    /  \   /   \   /   \
[S1][S2][S3][S4][S5][S6][S7][S8]
```

**Reply by nashqueue** [reply]
One of the verification steps of compact fraud proofs will be to see if the shares from the fraud proof exist in the blob that is referenced. And before that we presumably already verified the inclusion of the blob. 

The reason why you want to have compact FPs instead of full blocks is still the fact that a LN who is processing a FP for every block for the whole block will turn into a FN. Maybe that is feasible resources wise, but that is not something desirable.

**Reply by rootulp** [reply]
> is still the fact that a LN who is processing a FP for every block for the whole block will turn into a FN

Sorry for repeating myself but a LN that processes a valid FP for the first block will shut down. A LN that processes an invalid FP for the first block will block the FN that sent that FP. 

I don't see why FPs need to reference a subset of shares rather than the entire blob. Given this API is for the blob module, I'm hesitant to expose the raw share type (and associated fraud proofs for them) from this API. If there is a convincing use case that warrants a need to index directly to shares, we may consider a separate share API for it.

**Reply by musalbas** [reply]
It's because there is indeed "an assumption that the LN doesn't have hardware resource requirements to download S1 through S8".

Compact fraud proofs are just an optimization for DoS-resistance to prevent attackers from making LNs download too much junk data (i.e. many multi-megabyte invalid fraud proofs coming from different IP addresses).

**Comment by evan-forbes** [comment]
We also use the [share version](https://github.com/celestiaorg/celestia-core/blob/77a1de71feadebaa89c5ecaef0309fd8b9f2f921/types/block.go#L1076-L1078), which is added to the share and is needed to properly recreate any commitments. Currently, we only have a share version of 0, but we might want to consider adding it here given its importance.

ref [share specs](https://github.com/celestiaorg/celestia-app/blob/specs-staging/specs/src/specs/data_structures.md#share)

**Reply by rootulp** [reply]
+1, a user can't generate a share commitment without specifying a share version. Ref: https://github.com/rootulp/celestia-app/blob/d2eb6b0fd9f2b83859499ac315962ea7a67bd13f/x/blob/types/payforblob.go#L127-L133

In addition to the share version, a blob will likely also contain the new field `namespaceVersion` per https://github.com/celestiaorg/celestia-app/blob/6d27b78aa64a749a808e84ea682352b8b551fbd7/docs/architecture/adr-014-versioned-namespaces.md

**Reply by Wondertan** [reply]
Good point. Do we want the user to specify the share version? An alternative could be the version that celestia-node support and harcodes the version.

**Reply by rootulp** [reply]
Good question. I think user should have the option to specify `namespaceVersion` and `shareVersion` from the outset so that our APIs are capable of introducing new versions for these fields without a breaking change to the API.

If either of those fields aren't specified by the user, we can default to `0` b/c they will be the only supported `namespaceVersion` and `shareVersion` at mainnet launch.

**Reply by Wondertan** [reply]
Ok, to support the namespace version, we may do an additional namespace id type. This type will have versioning exposed, and users will be able to instantiate a namespace with the desired version. The namespace id type will then be used by our APIs.

For share versioning, it might be a bit trickier, as the blob module does not know about shares, but if it's really necessary we can have a share version on the Blob struct that signals to the node which version to use.

**Reply by Wondertan** [reply]
> 0 b/c they will be the only supported namespaceVersion and shareVersion at main net launch.

Still, I would like to have a future-proof version of the API. But the question is, once we introduce version e.g. 1, will there be any use case for users to use the previous version? I presume yes, but it could be that we only support versions to be able to parse older blocks as newer ones always use the new version.

**Reply by rootulp** [reply]
>  will there be any use case for users to use the previous version?

Yes because a new namespace version will be interpreted as a completely distinct namespace. In other words:

```go
namespaceVersion := 0
namespaceId := []byte{1, 1, 1, ...} // 32 bytes

namespaceVersion := 1
namespaceId := []byte{1, 1, 1, ...} // 32 bytes
```

will publish to two distinct namespaces so roll-ups must query by namespaceVersion + namespaceId. In order to avoid breaking roll-ups on mainnet, celestia-node can't assume which version they want to use after we introduce more than one. In fact, it's likely safest to retain the default of `namespaceVersion=0` after we introduce namespaceVersion 1 so that roll-ups must explicitly opt in to the new namespace version and therefore accept that they are migrating their roll-up's canonical namespace.

> that we only support versions to be able to parse older blocks as newer ones always use the new version.

If we disabled ability to use previous namespace versions, that would be a breaking change and at the same time we could make a breaking change to this API.

**Reply by Wondertan** [reply]
And I guess the same is true for the share versioning, correct?

**Comment by nashqueue** [comment]
Should 'Submit' receive [ ] of  Blobs? If not how do you submit multiple blobs under the same PFB? 

With this Interface, we would do this:

- Create Commitment of Data
- calculate commitment of Header
- Submit Header and Data with submit -> height
- Get Proof for Header and Data with height and commitment
- Do light clients verify inclusion by commitment? 
	
I thought inclusion was supposed to be verified by proof for now as we gossip it over rollkit p2p.

**Reply by Wondertan** [reply]
Submit receives a variadic Blob argument, which is effectively a slice.

**Reply by Wondertan** [reply]
See example: https://gobyexample.com/variadic-functions:

**Reply by Wondertan** [reply]
> I thought inclusion was supposed to be verified by proof for now as we gossip it over rollkit p2p.

You cannot verify inclusion just with Blob. You need additional context like DAHeader. The `Included` provides you this

**Reply by Wondertan** [reply]
Ah, I see the confusion. The Included does not have Proof as param. Fixed

**Reply by nashqueue** [reply]
LGTM

**Comment by distractedm1nd** [comment]
Idea: would be helpful to add a `Commitment` method on the interface that returns the commitment of the blob. This would be helpful for using the blob module over the RPC CLI, and would be a good intermediate step for client implementations that haven't ported the logic for calculating commitments yet.

**Reply by Wondertan** [reply]
The cli/cmd/client logic can compute the commitment locally. Not on the remote node.

**Reply by Wondertan** [reply]
The principle for the APIs we have is that they are statefull. Stateless operations that don't require any special state/functionality from the node can be and should be done outside of the API.

**Reply by musalbas** [reply]
What's the difference between that and this (from original post)?

```go
// Commitment computes the commitment of the Blob. 
// It is computed by splitting the blob into shares and building the Merkle subtree to be included after Submit.
func (b *Blob) Commitment() Commitment
```

**Reply by Wondertan** [reply]
Ryan proposed Commitment to be a method of BlobModule in addition to the method of the actual Blob(what you’ve shared)

**Comment by rootulp** [comment]
[non-blocking] I think it's a bit cumbersome that the following methods accept as an argument namespace and commitment. Given a commitment is sufficient to uniquely identify a blob, it seems a bit easier for a caller to only specify the commitment and not both the namespace and commitment. Given this is a UX concern and the caller likely knows the namespace they are querying for, this issue isn't blocking and could be considered a future improvement.

```diff
	// Get retrieves the blob by commitment. 
-	Get(context.Context, uint64, namespace.ID, Commitment) (Blob, error)
+	Get(context.Context, uint64, Commitment) (Blob, error) 

        // GetProof gets the Proof for a blob if it exists at the given height under the namespace. 
-        GetProof(context.Context, height uint64, namespace.ID, Commitment) (Proof, error)
+        GetProof(context.Context, height uint64, Commitment) (Proof, error)

	// Included checks whether a blob's given commitment(Merkle subtree root) is included at the given height under a particular namespace.
	// Useful for light clients wishing to check the inclusion of the data.
-	Included(context.Context, height uint64, namespace.ID, Commitment, Proof) (bool, error)
+ 	Included(context.Context, height uint64, Commitment, Proof) (bool, error)
```

**Reply by Wondertan** [reply]
Can we make commitment to include namespace?

**Reply by vgonkivs** [reply]
No, afaik it's not possible(currently)

**Reply by rootulp** [reply]
I think the share commitment could include the namespace. The last step of creating a share commitment is the step that loses the namespace field (i.e. generating a binary Merkle tree where the leaves of the tree are NMT subtree roots). So it seems possible for that step to prepend the namespace ID from the subtree roots (remember all the subtree roots will share a min/max namespace) to the resulting Merkle root. Note: this has the side-effect that the commitment size changes from 32 bytes to 65 bytes (1 byte for namespace version + 32 bytes for namespace ID + 32 bytes for original Merkle root).

**Comment by vgonkivs** [comment]
In order to verify the Proof, we will also need rowRoots(to verify nmt.Proofs against them), so I'd suggest extending BlobModule interface with the:
`Verify(Blob, Proof)(bool)`

**Reply by Wondertan** [reply]
Verify can be a Blob method as well, which takes DAH header as a param

**Reply by vgonkivs** [reply]
That's true, but then rollkit will need to get DAH first...

**Reply by Wondertan** [reply]
Why do we need this as separate method instead of sub operation of Included?

**Reply by vgonkivs** [reply]
To ensure that the Proof corresponds to the Blob.

**Reply by vgonkivs** [reply]
If it's not needed, I'm ok with sub-operation inside Included

**Reply by Wondertan** [reply]
I am not sure we need this a "sub operation" as well. 

Included needs to verify inclusion of the commitment, not the blob(that's an important subtlety we found during on-site).

Included doesn't need to know the blob

**Comment by evan-forbes** [comment]
leaving a comment about paying for gas fees when submitting blobs here after discussing with @vgonkivs:

We likely want the user to have some input that determines the price of gas. If we have that, then we can estimate the gas limit, and then determine the fee. Both of those are then signed over. This simplifies the api, because the user only enters one thing, instead of fee and gas limit, which are both difficult to calculate.

Determining gas price could be as simple as the user specifying a direct price, or more complex where we estimate what the lowest gas price is to achieve some outcome, such as getting included in the next block, or getting included in any of the next 10 blocks etc.

**Reply by nashqueue** [reply]
How can you encompass the fee and gaslimit in one parameter? 
It feels like estimation should be easy as you can simulate sparse shares construction and calculate the resulting cost. 
Maybe also just a formula like gas(bytes) = gas_per_share * shares(bytes) and 
shares(bytes)= roundup [ |bytes - (512 - first_share_overhead )| / (512 - share_overhead)]
As I understand, we only support one signature type right now? So Transactions and PFB Transactions should be fixed. The only variable is how many blobs are in one pfb transaction. I bet I am missing something here like multi Transactions. 


Another fun idea: Make a TWAP for Gas.

**Reply by cmwaters** [reply]
There's a simulateTx endpoint on consensus nodes that could be used to calculate the gasLimit. As I mentioned elsewhere, the problem with using an offline formula is that the `gas_per_blob_byte` is a parameter that could change via governance. I think the ideal solution would be to have gas refunds (the excess gas goes back to your account). 

As to gas prices, we could start by using the `DefaultMinGasPrice` set in `celestia-app`. In the future it would be nice to have it as an endpoint that is exposed (or perhaps returned in the `SimulateTx` call) since the value could fluctuate.
